### PR TITLE
[SPMD][PoC] Tensor sharding util

### PR DESCRIPTION
**Pull request by yeounoh**
This is taken from #3684 and addresses #3871 . The original PR will be broken down into 3 PRs, and this will be the first PR to land.

The changes include:
- Add at::Tensor sharding util & sharding tests.
- Refactor XLAShardedTensor & unit tests.

### CODE REVIEW

File: torch_xla/experimental/xla_sharded_tensor.py [hunk 1]
```diff
@@ -32,6 +32,76 @@ class XLAShardedTensor(torch.Tensor):
...
 
+  # XLAShardedTensor behaves like a unpartitioned,
+  # combined tensor on the host machine. When user annotates,
+  # this is simply set to the input tensor. When an XLA partitioned
+  # output tensor returns (or sharding propagated intermediate tensors)
+  # as XLAShardedTensor, the backend gathers global data across devices
+  # and materialize and set `global_tensor` on the host; the actual device
+  # data still remain on individual device as sharded or replicated.
+  # Note: we should drop this reference, and force all gather on each access.
+  global_tensor: torch.Tensor
```
Comment (JackCaoG) [inline torch_xla/experimental/xla_sharded_tensor.py:43]: Are these fields and this class tested? It seems pretty random in this pr.

**Review by JackCaoG — CHANGES_REQUESTED** [review]
Mostly lgtm, slightly worried about the test coverage.

File: torch_xla/experimental/xla_sharded_tensor.py (hunk 1 above)
Comment (yeounoh) [reply torch_xla/experimental/xla_sharded_tensor.py:43]: It's part of the first separate PR -- once we land it, rebase and this will be gone.

File: torch_xla/csrc/xla_sharding_util.h [hunk 2]
```diff
@@ -13,14 +14,22 @@ class ShardingUtil {
...
   static xla::HloModuleProto SpmdPartitioningPass(
       const xla::HloModuleProto& hlo_proto,
       bool conv_halo_exchange_always_on_lhs = true,
       bool choose_faster_windowed_einsum_over_mem = false,
       bool unroll_windowed_einsum = false,
       bool bidirectional_windowed_einsum = false);
+
+  // Shard a tensor and returns the sharded tensors based on the `sharding`
+  // spec.
+  static std::vector<at::Tensor> ShardTensor(
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.h:30]: It seems that we use `tensor` often as variable names referring to `at::Tensor` in the code, and `xtensor` for `XLATensor`. `ShardAtTensor` sounds like "shard at tensor" (?!), and don't like the device name in the function, as it's device agnostic. I think `ShardingUtil::ShardTensor` (it's static for that reason) will be distinctive enough throughout the codebase without confusion. I think we could change it to `ShardSourceTensor`, I've seen a name source tensor referring to `at::Tensor` in our code.

File: torch_xla/csrc/xla_sharding_util.cpp [hunk 3]
```diff
@@ -70,21 +66,70 @@ xla::HloModuleProto ShardingUtil::SpmdPartitioningPass(
...
+std::vector<at::Tensor> ShardingUtil::ShardTensor(
+    const at::Tensor& tensor, const xla::OpSharding sharding,
+    const std::vector<std::string>& devices) {
+  TF_LOG(INFO) << "ShardTensor with sharding type(" << sharding.type() << ")..."
+               << std::endl;
+  std::vector<at::Tensor> shards(devices.size());
+  if (sharding.type() == xla::OpSharding::REPLICATED) {
+    std::fill_n(shards.begin(), shards.size(), tensor);
+  } else if (sharding.type() == xla::OpSharding::OTHER) {
+    CHECK_EQ(devices.size(), sharding.tile_assignment_devices().size())
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:98]: +1 ack.

File: torch_xla/csrc/xla_sharding_util.cpp [hunk 4]
```diff
@@ -70,21 +66,70 @@ xla::HloModuleProto ShardingUtil::SpmdPartitioningPass(
...
-      /*num_partitions=*/num_partitions, /*num_replicas=*/num_replicas);
-
   xla::HloPassPipeline pass("spmd-partitioning");
   pass.AddPass<xla::HloVerifier>(/*layout_sensitive=*/false,
                                  /*allow_mixed_precision=*/false);
-  pass.AddPass<xla::ShardingPropagation>(/*is_spmd=*/true);
-  pass.AddPass<xla::spmd::SpmdPartitioner>(/*num_partitions=*/num_partitions,
-                                           /*num_replicas=*/num_replicas,
-                                           options, collective_ops_creator);
+  // TODO(yeounoh) side-effecting ops gets assigned replicated sharding.
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:72]: The large PR already parameterized those numbers (which will land after landing the first 2 PRs). As per the context, this function is mainly used for debugging -- let me make the header comment more explicit.

File: torch_xla/csrc/xla_sharding_util.cpp [hunk 5]
```diff
@@ -19,29 +23,21 @@ using xla::internal::XlaBuilderFriend;
...
-      {
-        // Annotate the full-shape input with the sharding.
-        xla::XlaScopedShardingAssignment assign_sharding(
-            lowering_ctx->builder(), *xla_node->GetSharding());
-        xla::CustomCall(lowering_ctx->builder(),
-                        /*call_target_name=*/"Sharding",
-                        xla_node->Lower(lowering_ctx), xla_node->xla_shape(),
-                        /*opaque=*/"");
-      }
+      *instruction->mutable_sharding() = *xla_node->GetSharding();
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:34]: Good point. I don't think it's part of the node hash -- let me address this separately. in the final PR.

[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:34]: Also, adding a separate IR node for sharding is tricky as it breaks the 1 to 1 relationship between IR node and HLO inst, the compiler consumes/removes the `Node -> XlaOp`, and annotate the previous instruction (e.g., `*instruction->mutable_sharding()`) itself.

File: torch_xla/csrc/xla_sharding_util.cpp [hunk 6]
```diff
@@ -70,21 +66,70 @@ xla::HloModuleProto ShardingUtil::SpmdPartitioningPass(
...
+  std::vector<at::Tensor> shards(devices.size());
+  if (sharding.type() == xla::OpSharding::REPLICATED) {
+    std::fill_n(shards.begin(), shards.size(), tensor);
+  } else if (sharding.type() == xla::OpSharding::OTHER) {
+    CHECK_EQ(devices.size(), sharding.tile_assignment_devices().size())
+        << "Invalid sharding tile_assignment_devices.size(): expected "
+        << devices.size() << ", actual "
+        << sharding.tile_assignment_devices().size();
+    CHECK(sharding.tile_shape().dimensions_size() <= 2);
+    CHECK(tensor.sizes().size() >= sharding.tile_shape().dimensions_size());
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:103]: +1 ack.

File: test/test_xla_sharding.py [hunk 7]
```diff
@@ -32,12 +30,27 @@ def test_xla_sharded_tensor(self):
...
+    # assignments.
+
+  def test_mark_sharding(self):
+    t1 = torch.randn(1, 128, device='cpu')
+    t2 = torch.randn(1, 128, device='cpu')
+    expected = t1 @ t2.T
+
+    xt1 = t1.to(xm.xla_device())
+    xt2 = t2.to(xm.xla_device())
+    n_devices = xm.xrt_world_size()
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply test/test_xla_sharding.py:43]: This will be 1 or 2 depending on the setup. SPMD is not available on GPU. I think we should start a separate workstream to support GPU -- the device assignment et al would need more work.

File: test/cpp/test_xla_sharding.cpp [hunk 8]
```diff
@@ -1,59 +1,67 @@
...
+  tensor = at::ones({8, 7, 4}, at::TensorOptions(at::kFloat));
+  xla::Array2D<int64_t> mesh({
+      {0, 1, 2, 3},
+      {4, 5, 6, 7},
+  });
+  sharding = xla::HloSharding::Tile(mesh).ToProto();
+  shards = ShardingUtil::ShardTensor(tensor, sharding, devices);
+  EXPECT_EQ(shards.size(), 8);
+  EXPECT_EQ(shards[0].sizes(), c10::ArrayRef<long>({4, 2, 4}));
+  EXPECT_EQ(shards[7].sizes(), c10::ArrayRef<long>({4, 1, 4}));
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply test/cpp/test_xla_sharding.cpp:57]: Regarding 1) let me add some more "concise" description to the method header. For 2), the argument (tensor) should be sharded without padding -- so the compiler can decide what to do with it (e.g., padding, reshaping).

File: test/cpp/test_xla_sharding.cpp [hunk 9]
```diff
@@ -1,59 +1,67 @@
...
-#include "torch_xla/csrc/torch_util.h"
+#include "torch_xla/csrc/xla_sharding_util.h"
 #include "torch_xla_test.h"
 
 namespace torch_xla {
 namespace cpp_test {
-
-// A placeholder for future SPMD sharding tests.
+namespace {
+bool TwoEqualXlaData(torch::lazy::BackendDataPtr a,
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply test/cpp/test_xla_sharding.cpp:21]: +1 Ack, sounds good lol

**Comment by yeounoh** [comment]
> Mostly lgtm, slightly worried about the test coverage.
Yea, the unit tests cover individual components on the code paths for different paths/branches. We would need to run more extensive integration testing outside (manually) the "mark_sharding_test". Eventually, we should add a nightly test that uses all the basic SPMD components.

File: torch_xla/experimental/xla_sharded_tensor.py (hunk 1 above)
Comment (yeounoh) [reply torch_xla/experimental/xla_sharded_tensor.py:43]: @JackCaoG correction, this is the first PR. And yes, this is random and wrong -- the same code is copied over twice lol Everything, except "Not implemented" methods are tested -- once we land this, Steven will add more extensive unit tests around, too.

File: torch_xla/csrc/xla_sharding_util.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:72]: https://github.com/pytorch/xla/blob/e05452b37da02409d11c5270332e539a7b5d36b2/torch_xla/csrc/xla_sharding_util.h#L25 cc @JackCaoG

File: test/cpp/test_xla_sharding.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply test/cpp/test_xla_sharding.cpp:57]: https://github.com/pytorch/xla/blob/504f0c881177e587dfd0cc6ddf79000c82374293/torch_xla/csrc/xla_sharding_util.h#L28

File: test/test_xla_sharding.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply test/test_xla_sharding.py:46]: Added 
```
@unittest.skipIf(
      len(xm.get_xla_supported_devices("GPU")) > 0,
      "SPMD is currently not supported on GPU.")
      ```

[the remark this answers is not in the document]
Comment (yeounoh) [reply test/test_xla_sharding.py:46]: Created a bug to track GPU enablement.

File: torch_xla/csrc/xla_sharding_util.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:34]: Created a bug to track this.

File: torch_xla/csrc/xla_sharding_util.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp]: Done

File: torch_xla/csrc/xla_sharding_util.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp]: Done

File: test/cpp/test_xla_sharding.cpp (hunk 9 above)
Comment (JackCaoG) [inline test/cpp/test_xla_sharding.cpp:21]: ditto

[the remark this answers is not in the document]
Comment (yeounoh) [reply test/cpp/test_xla_sharding.cpp]: Done, just now. THis function will actually be used in the next PR.

Comment (yeounoh) [reply test/cpp/test_xla_sharding.cpp]: Sorry for missing this. Just addressed this.

Comment (yeounoh) [reply test/cpp/test_xla_sharding.cpp]: Done.

File: test/test_xla_sharding.py [hunk 10]
```diff
@@ -32,12 +30,30 @@ def test_xla_sharded_tensor(self):
...
-    # TODO(yeounoh) sharding_spec is broken. The intention is to
-    # check if the returned sharding spec holds the correct device
-    # assignment as intended.
-    # assert device_assignment == t1_sharded.sharding_spec[0]
+    # TODO(yeounoh) Check if the returned sharding spec holds the correct device
+    # assignments.
+
+  @unittest.skipIf(
+      len(xm.get_xla_supported_devices("GPU")) > 0,
+      "SPMD is currently not supported on GPU.")
```
Comment (JackCaoG) [inline test/test_xla_sharding.py:38]: I think it is supported on GPU, but we don't need to support it in first version.

File: test/test_xla_sharding.py [hunk 11]
```diff
@@ -1,26 +1,24 @@
...
 import torch_xla.utils.utils as xu
 import torch_xla.experimental.xla_sharding as xs
 from torch_xla.experimental.xla_sharded_tensor import XLAShardedTensor
-import unittest
-import numpy as np
-
-# A placeholder for future sharding tests
 
 
 class XlaShardingTest(unittest.TestCase):
```
Comment (JackCaoG) [inline test/test_xla_sharding.py:16]: can we add a assert on `PJRT_DEVICE` somewhere in this file, we won't support SPMD on xrt.

File: torch_xla/csrc/xla_sharding_util.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (JackCaoG) [reply torch_xla/csrc/xla_sharding_util.cpp:34]: I don't fully get comment above. I think we have to create new node for hash because 

```
t1 = torch.tensor(1)
t2 = t1 + 3
xm.mark_shard(t1, something)
```

but this time we already compute the IR for t2. in above case we need to make sure `mark_shard` creates a new IR and new tensor.

through now I think about it this approach also has issues, we don't want to actually execute the t1 without shards. Ideally, we should not do any operation on tensor on virtual devices that does not have shard info. Let's talk more offline I guess

File: test/test_xla_sharding.py (hunk 10 above)
Comment (yeounoh) [reply test/test_xla_sharding.py:38]: +1 rephrasing the comment to avoid confusion. I really think we should focus on the TPU (and CPU) first. Once we hit all the milestones, adding GPU support should be more manageable/easier.

File: torch_xla/csrc/xla_sharding_util.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:34]: The example is a bit confusing to me... do we want to compute t2 with or without sharding. Just like we will be deferring backend data creation with virtual device and initiate only sharded data transfer with mark_sharding, could we update the hash with sharding annotation? We are annotating the tensor and thus its IR node. We don't compute or haven't generated HLO at the annotation time, yet.

[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:34]: > Created a bug to track this.

Synced with @JackCaoG , we will follow up in a separate PR -- this is not a blocker for the ongoing PoC implementation and testing, but critical for us to get done.

File: test/test_xla_sharding.py (hunk 11 above)
Comment (yeounoh) [reply test/test_xla_sharding.py:18]: +1, though I prefer skip_test over assertion failure. I will add this to the testing conditions.

File: test/test_xla_sharding.py (hunk 10 above)
Comment (yeounoh) [reply test/test_xla_sharding.py]: Done

File: test/test_xla_sharding.py (hunk 11 above)
Comment (yeounoh) [reply test/test_xla_sharding.py:20]: Done

File: torch_xla/experimental/xla_sharding.py [hunk 12]
```diff
@@ -30,46 +26,44 @@ def mark_sharding(t: Union[torch.Tensor,
...
     replicated = True
+  elif any(d is None for d in partition_spec):
+    partial = True
+
+  # TODO(yeounoh) suport partially replicated sharding.
+  assert not partial, "Partial replication is currently not supported."
 
   if isinstance(t, XLAShardedTensor):
-    # Update sharding annotation
     torch_xla._XLAC._xla_mark_sharding(t.global_tensor, tile_assignment,
```
Comment (JackCaoG) [inline torch_xla/experimental/xla_sharding.py:65]: per our discussion, we might want to error out here and ask user to manully clear sharding which will move data back to cpu.

**Review by JackCaoG — APPROVED** [review]
LGTM, thanks!
